### ISSUE TITLE
Badly placed 'Unexpected end of macro invocation' when calling a macro with no arguments

### ISSUE THREAD

**Issue by thepowersgang** [rust-lang/rust#26615]
When a  macro that expects arguments is not passed any, the "unexpected end of macro invocation" message is placed at the start of the crate root file, instead of the invocation.

``` rust
// Test case

macro_rules! mac {
    ($e:expr) => ("hello");
}

fn main() {
    println!("{}", mac!());
}
```

```
<anon>:1:1: 1:1 error: unexpected end of macro invocation
<anon>:1 // Test case
         ^
playpen: application terminated with error code 101
```

**Comment by thepowersgang** [comment]
Mostly fixed in #27584, above test case still produces a slightly unexpected result on playpen nightly, but points to the correct line (just to the `println!` invocation instead of `mac!`)

```
<anon>:8:5: 8:28 error: unexpected end of macro invocation
<anon>:8     println!("{}", mac!());
             ^~~~~~~~~~~~~~~~~~~~~~~
```

**Comment by vberger** [comment]
There are still some pathological cases, for example with this file:
```rust
fn main() {}

fn foo() {
    for _ in 0..3 {
        /*
            some
            code
            here
        */
        println!();
        /*
            some
            code
            here
        */
    }
}
```
The generated error (on nigthly playpen) is:
```
<anon>:4:5: 16:6 error: unexpected end of macro invocation
<anon>:4     for _ in 0..3 {
<anon>:5         /*
<anon>:6             some
<anon>:7             code
<anon>:8             here
<anon>:9         */
         ...
```

**Comment by TimNN** [comment]
Both of theses examples probably have the same root cause: The "unexpected end of macro invocation" error apparently always reports the span of the outermost ast node which requires expansion. See [this gist](https://gist.github.com/9229a43c3d5ae0e200dc) ([playground](https://play.rust-lang.org/?gist=9229a43c3d5ae0e200dc&version=nightly)) for some examples of which span is reported in different contexts.

**Comment by TimNN** [comment]
The situation seems to have improved in the latest nightly: now only nested macro invocations have an incorrect spam and not macro invocations inside of other expanded things. Updated [gist](https://gist.github.com/62c717f26ad1b183cb94) ([playground](https://play.rust-lang.org/?gist=62c717f26ad1b183cb94&version=nightly)).

**Comment by TimNN** [comment]
Looking over some pr's since my last comment, the improvement probably happened as part of #28857.

**Comment by Mark-Simulacrum** [comment]
As noted above, non-nested macro invocations are now fixed, and those are tracked separately (#33871). Closing.
